### PR TITLE
Fix 'type annotations needed' error with opaque types

### PR DESCRIPTION
**Pull request by Aaron1011**
Related: #66426

This commit adds handling for opaque types during inference variable
fallback. Type variables generated from the instantiation of opaque
types now fallback to the opaque type itself.

Normally, the type variable for an instantiated opaque type is either
unified with the concrete type, or with the opaque type itself (e.g when
a function returns an opaque type by calling another function).

However, it's possible for the type variable to be left completely
unconstrained. This can occur in code like this:

```rust
pub type Foo = impl Copy;
fn produce() -> Option<Foo> {
    None
}
```

Here, we'll instantatiate the `Foo` in `Option<Foo>` to a fresh type
variable, but we will never unify it with anything due to the fact
that we return a `None`.

This results in the error message:
```
type annotations needed: cannot resolve `_: std::marker::Copy
```

pointing at `pub type Foo = impl Copy`.

This message is not only confusing, it's incorrect. When an opaque type
inference variable is completely unconstrained, we can always fall back
to using the opaque type itself. This effectively turns that particular
use of the opaque type into a non-defining use, even if it appears in a
defining scope.

### CODE REVIEW

File: src/test/ui/type-alias-impl-trait/fallback.rs [hunk 1]
```diff
@@ -0,0 +1,22 @@
+// Tests that we correctly handle the instantiated
+// inference variable being completely unconstrained.
+//
+// check-pass
```
Comment (Centril) [inline src/test/ui/type-alias-impl-trait/fallback.rs]: My understanding of the RFC (at least the spirit of it, it doesn't elaborate on cases like these, cc @cramertj) is that each function fully determines each referenced opaque type. In this case `_make_iter` does fully determine the opaque type `Foo`. However, `_produce`, without consulting other functions (imagine `_make_iter` did not exist), can only determine that `Wrapper::Second: Wrapper<?opaque_infer_var_0>`.

Other notable examples which we may consult are (the first of which you included as a snippet in your code):
```rust
fn alpha() -> Option<impl Copy> { None }
//~^ ERROR type annotations needed: cannot resolve `_: std::marker::Copy`
```
and:
```rust
fn main() {
    let beta = None;
    //~^ ERROR type annotations needed for `std::option::Option<T>`
}
```

In conclusion, while this is behavior that may be desirable, I don't think it follows from https://github.com/rust-lang/rfcs/pull/2071's notion of each function independently determining opaque types, and so this test shouldn't be `// check-pass`.

**Comment by Centril** [comment]
r? @varkor 
cc @nikomatsakis

File: src/test/ui/type-alias-impl-trait/fallback.rs (hunk 1 above)
Comment (Aaron1011) [reply src/test/ui/type-alias-impl-trait/fallback.rs]: @Centril: I believe the RFC explicitly alows this behavior:

 > A body or initializer must either fully constrain or place no constraints upon a given existential type.

See the `return_foo` example in the RFC.

Comment (Centril) [reply src/test/ui/type-alias-impl-trait/fallback.rs]: Hmm; that does seem to follow yes. Can you amend the test with the rationale given here as a comment?

Comment (Aaron1011) [reply src/test/ui/type-alias-impl-trait/fallback.rs]: Updated

File: src/test/ui/type-alias-impl-trait/fallback.rs [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+    First(T),
+    Second
+}
+
+fn _make_iter() -> Foo {
+    true
+}
+
+fn _produce() -> Wrapper<Foo> {
+    Wrapper::Second
```
[the remark this answers is not in the document]
Comment (Centril) [reply src/test/ui/type-alias-impl-trait/fallback.rs]: Final type is `bool` as determined by `constrained_foo`.

File: src/librustc_typeck/check/mod.rs [hunk 3]
```diff
@@ -231,6 +231,13 @@ pub struct Inherited<'a, 'tcx> {
     // 'de-opaque' OpaqueTypeDecl, after typeck is done with all functions.
     opaque_types: RefCell<DefIdMap<OpaqueTypeDecl<'tcx>>>,
 
+    /// A map from inference variables created from opaque
+    /// type instantiations (ty::Infer) to the actual opaque
```
Comment (Centril) [inline src/librustc_typeck/check/mod.rs]:
```suggestion
    /// type instantiations (`ty::Infer`) to the actual opaque
```

File: src/librustc_typeck/check/mod.rs [hunk 4]
```diff
@@ -231,6 +231,13 @@ pub struct Inherited<'a, 'tcx> {
...
+    /// type (ty::Opaque). Used during fallback to map unconstrained
```
Comment (Centril) [inline src/librustc_typeck/check/mod.rs]:
```suggestion
    /// type (`ty::Opaque`). Used during fallback to map unconstrained
```

File: src/librustc/infer/opaque_types/mod.rs [hunk 5]
```diff
@@ -24,6 +24,10 @@ pub type OpaqueTypeMap<'tcx> = DefIdMap<OpaqueTypeDecl<'tcx>>;
 /// appear in the return type).
 #[derive(Copy, Clone, Debug)]
 pub struct OpaqueTypeDecl<'tcx> {
+
+    /// The opaque type (`ty::Opaque`) for this declaration
```
Comment (Centril) [inline src/librustc/infer/opaque_types/mod.rs]:
```suggestion
    /// The opaque type (`ty::Opaque`) for this declaration.
```

File: src/librustc_typeck/check/mod.rs [hunk 6]
```diff
@@ -231,6 +231,13 @@ pub struct Inherited<'a, 'tcx> {
...
+    /// opaque type inference variables to their corresponding
+    /// opaque type
```
Comment (Centril) [inline src/librustc_typeck/check/mod.rs]:
```suggestion
    /// opaque type.
```

File: src/librustc_typeck/check/mod.rs [hunk 7]
```diff
@@ -937,9 +945,46 @@ fn typeck_tables_of(tcx: TyCtxt<'_>, def_id: DefId) -> &ty::TypeckTables<'_> {
...
         let mut fallback_has_occurred = false;
+
+        // We do fallback in two passes, to try to generate
+        // better error messages.
+        // The first time, we do *not* replace opaque types.
+        for ty in &fcx.unsolved_variables() {
+            fallback_has_occurred |= fcx.fallback_if_possible(ty, false /* opaque_fallback */);
+        }
+        // We now see if we can make progress. This might
+        // cause us to unify inference varaibles for opaque types,
```
Comment (Centril) [inline src/librustc_typeck/check/mod.rs]:
```suggestion
        // cause us to unify inference variables for opaque types,
```

File: src/librustc_typeck/check/mod.rs [hunk 8]
```diff
@@ -937,9 +945,46 @@ fn typeck_tables_of(tcx: TyCtxt<'_>, def_id: DefId) -> &ty::TypeckTables<'_> {
...
+        //
+        // ```rust
+        // type MyType = impl Copy;
+        // fn produce() -> MyType { true }
+        // fn bad_produce() -> MyType { panic!() }
+        // ```
+        //
+        // we want to unify the opaque inference variable in `bad_produce`
+        // with the diverging fallback for `panic!` (e.g. `()` or `!`),
+        // This will produce a nice error messaeg about conflicting concrete
```
Comment (Centril) [inline src/librustc_typeck/check/mod.rs]:
```suggestion
        // This will produce a nice error message about conflicting concrete
```

**Review by Centril — COMMENTED** [review]
nits

File: src/test/ui/impl-trait/where-allowed-2.stderr [hunk 9]
```diff
@@ -0,0 +1,11 @@
+error[E0720]: opaque type expands to a recursive type
+  --> $DIR/where-allowed-2.rs:6:30
+   |
+LL | fn in_adt_in_return() -> Vec<impl Debug> { panic!() }
+   |                              ^^^^^^^^^^ expands to a recursive type
+   |
+   = note: type resolves to itself
```
[the remark this answers is not in the document]
Comment (Aaron1011) [reply src/test/ui/impl-trait/where-allowed-2.stderr]: Yeah, neither the new error nor the old error is very helpful.

File: src/librustc_typeck/check/mod.rs [hunk 10]
```diff
@@ -937,9 +945,46 @@ fn typeck_tables_of(tcx: TyCtxt<'_>, def_id: DefId) -> &ty::TypeckTables<'_> {
...
+        // we will generate a confusing type-check error that does not explicitly
+        // refer to opaque types.
+        fcx.select_obligations_where_possible(fallback_has_occurred, |_| {});
+
+        // We now run fallback again, but this time we allow it to replace
+        // unconstrained opaque type variables, in addition to performing
+        // other kinds of fallback.
         for ty in &fcx.unsolved_variables() {
-            fallback_has_occurred |= fcx.fallback_if_possible(ty);
+            fallback_has_occurred |= fcx.fallback_if_possible(ty, true /* opaque_fallback */);
```
[the remark this answers is not in the document]
Comment (Aaron1011) [reply src/librustc_typeck/check/mod.rs]: Done

**Comment by Aaron1011** [comment]
@Centril @estebank: I've addressed your comments.

File: src/librustc_typeck/check/mod.rs [hunk 11]
```diff
@@ -937,9 +945,46 @@ fn typeck_tables_of(tcx: TyCtxt<'_>, def_id: DefId) -> &ty::TypeckTables<'_> {
...
+        // In code like this:
+        //
+        // ```rust
+        // type MyType = impl Copy;
+        // fn produce() -> MyType { true }
+        // fn bad_produce() -> MyType { panic!() }
+        // ```
+        //
+        // we want to unify the opaque inference variable in `bad_produce`
+        // with the diverging fallback for `panic!` (e.g. `()` or `!`),
```
Comment (varkor) [inline src/librustc_typeck/check/mod.rs]:
```suggestion
        // with the diverging fallback for `panic!` (e.g. `()` or `!`).
```

File: src/librustc_typeck/check/mod.rs [hunk 12]
```diff
@@ -937,9 +945,46 @@ fn typeck_tables_of(tcx: TyCtxt<'_>, def_id: DefId) -> &ty::TypeckTables<'_> {
...
+        fcx.select_obligations_where_possible(fallback_has_occurred, |_| {});
+
+        // We now run fallback again, but this time we allow it to replace
+        // unconstrained opaque type variables, in addition to performing
+        // other kinds of fallback.
+        for ty in &fcx.unsolved_variables() {
+            fallback_has_occurred |= fcx.fallback_if_possible(ty, FallbackMode::All);
         }
+
+        // See if we can make any more progress
```
Comment (varkor) [inline src/librustc_typeck/check/mod.rs]:
```suggestion
        // See if we can make any more progress.
```

File: src/librustc_typeck/check/mod.rs [hunk 13]
```diff
@@ -2499,6 +2544,16 @@ enum TupleArgumentsFlag {
     TupleArguments,
 }
 
+/// Controls how we perform fallback for unconstrained
+/// type variables
```
Comment (varkor) [inline src/librustc_typeck/check/mod.rs]:
```suggestion
/// type variables.
```

File: src/librustc_typeck/check/mod.rs [hunk 14]
```diff
@@ -2499,6 +2544,16 @@ enum TupleArgumentsFlag {
...
+enum FallbackMode {
+    /// Do not fallback type variables to opaque types
```
Comment (varkor) [inline src/librustc_typeck/check/mod.rs]:
```suggestion
    /// Do not fallback type variables to opaque types.
```

File: src/librustc_typeck/check/mod.rs [hunk 15]
```diff
@@ -2499,6 +2544,16 @@ enum TupleArgumentsFlag {
...
+    NoOpaque,
+    /// Perform all possible kinds of fallback, including
+    /// turning type variables to opaque types
```
Comment (varkor) [inline src/librustc_typeck/check/mod.rs]:
```suggestion
    /// turning type variables to opaque types.
```

File: src/librustc_typeck/check/mod.rs [hunk 16]
```diff
@@ -2499,6 +2544,16 @@ enum TupleArgumentsFlag {
...
+    All
```
Comment (varkor) [inline src/librustc_typeck/check/mod.rs]:
```suggestion
    All,
```

File: src/test/ui/impl-trait/where-allowed-2.stderr (hunk 9 above)
[the remark this answers is not in the document]
Comment (varkor) [reply src/test/ui/impl-trait/where-allowed-2.stderr]: @Aaron1011: if there's not an existing issue open for this case, could you open one, so we don't forget about it?

**Review by varkor — COMMENTED** [review]
I'm happy with the implementation once the nits have been addressed and rebased into the main commits.

File: src/test/ui/impl-trait/where-allowed-2.stderr (hunk 9 above)
[the remark this answers is not in the document]
Comment (Aaron1011) [reply src/test/ui/impl-trait/where-allowed-2.stderr]: I've opened https://github.com/rust-lang/rust/issues/66523

**Comment by Aaron1011** [comment]
@varkor: Squashed
